src/autopas/utils/SimilarityFunctions.h

```diff
@@ -65,10 +65,41 @@ std::pair<double, double> calculateHomogeneityAndMaxDensity(const ParticleContai
   for (auto particleItr = container.begin(autopas::IteratorBehavior::owned); particleItr.isValid(); ++particleItr) {
     const auto &particleLocation = particleItr->getR();
 
-    const auto binIndex3d =
+    const auto binIndex3dUnsafe =
         autopas::utils::ArrayMath::floorToInt((particleLocation - container.getBoxMin()) / binDimensions);
+
+    // It is possible that floating point errors result in out of bounds indices.
+    // e.g. if there are 7 bins in the x dimension, and that particle is close to the right domain boundary, the
+    // division above might result in 7.0, which gets floored to 7 corresponding to a bin index that is out of bounds!
+
+    // We therefore check for particle indices that are out of bounds and, if the particle is close to the boundary we
+    // adjust the index. We don't care about floating point errors causing incorrect indices internally in the
+    // domain as, for particles so close to a boundary, it is somewhat arbitrary which bin they fall into.
+
+    const auto binIndex3DSafe = [&]() {
+      auto newBinIndex3D = binIndex3dUnsafe;
+      for (int i = 0; i < 3; ++i) {
```

Comment:
I find the body of the loop a bit repetitive and confusing to read. 
Here is my suggestion how you could refactor it: 

for each dimension 
 - Check that the particle is indeed within the container and throw an exception if not. (one if, exception call is only implemented once) 
 - use `std::clamp` to force the index into the bounds.

Reply:
See a6952e7d